crates/valence_command/src/parsers.rs

```diff
@@ -0,0 +1,327 @@
+    pub fn peek_n(&self, n: usize) -> &'a str {
+        self.0
+            .char_indices()
+            .nth(n)
+            .map(|(idx, _)| &self.0[..idx])
+            .unwrap_or(self.0)
+    }
+
+    /// Returns the next word without advancing the input
+    pub fn peek_word(&self) -> &'a str {
+        self.0
+            .char_indices()
+            .find(|(_, c)| c.is_whitespace())
+            .map(|(idx, _)| &self.0[..idx])
+            .unwrap_or(self.0)
+    }
+
+    /// Checks if the input is empty
+    pub fn is_done(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns the next character and advances the input
+    pub fn pop(&mut self) -> Option<char> {
+        let c = self.peek()?;
+        self.advance();
+        Some(c)
+    }
+
+    /// Returns the next n characters and advances the input
+    pub fn pop_n(&mut self, n: usize) -> &str {
+        let s = self.peek_n(n);
+        self.advance_n_bytes(s.len());
+        s
+    }
+
+    /// Returns the next word and advances the input
+    pub fn pop_word(&mut self) -> &str {
+        let s = self.peek_word();
+        self.advance_n_bytes(s.len());
+        s
+    }
+
+    /// Returns the rest of the input and advances the input
+    pub fn pop_all(&mut self) -> Option<&str> {
+        let s = self.0;
+        self.advance_n_bytes(self.0.len());
+        Some(s)
+    }
+
+    /// Returns the next word and advances the input
+    pub fn pop_to_next(&mut self, c: char) -> Option<&str> {
+        let pos = self.0.find(c)?;
+        let s = &self.0[..pos];
+        self.advance_n_bytes(pos);
+        Some(s)
+    }
+
+    /// Matches the case-insensitive string and advances the input if it matches
+    pub fn match_next(&mut self, string: &str) -> bool {
+        if self
+            .0
+            .to_lowercase()
+            .starts_with(string.to_lowercase().as_str())
+        {
+            self.advance_n_bytes(string.len());
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Skip all whitespace at the front of the input
+    pub fn skip_whitespace(&mut self) {
+        while let Some(c) = self.peek() {
+            if c.is_whitespace() {
+                self.advance();
+            } else {
+                break;
+            }
+        }
+    }
+
+    /// Set the inner string
+    pub fn into_inner(self) -> &'a str {
+        self.0
+    }
+
+    /// Get the difference between two strings from left to right
+    ///
+    /// # Safety
+    /// `other` must be a slice of the same string as `self`
+    pub unsafe fn get_diff(&self, other: &str) -> &str {
+        let mut far_ptr = self.0.as_ptr();
+        let mut near_ptr = other.as_ptr();
+
+        if near_ptr > far_ptr {
+            std::mem::swap(&mut near_ptr, &mut far_ptr);
+        }
+        let new_len = far_ptr.offset_from(near_ptr) as usize;
+
+        std::str::from_utf8_unchecked(std::slice::from_raw_parts(near_ptr, new_len))
+    }
```

Comment:
Not a fan of the unsafe code here. I'm also not confident its use is entirely sound.

Reply:
It is either this or storing some kind of cursor or traversed chars I'm afraid. The only unsafely is `far_ptr.offset_from` and `from_utf8_unchecked` as long as both slices are from the same string it should be completely fine.